File: python/src/autobi/dataset_builder.py
from pathlib import Path
from typing import Self, List
from tempfile import TemporaryDirectory

import pandas as pd
from py4j.java_gateway import Py4JJavaError

from autobi.core import AutobiJVM, takes, to_resolved_path_str


class DatasetBuilder:
    def __init__(self, jvm: AutobiJVM, params: str):
        try:
            self._jvm = jvm
            self._object = jvm._jvm.edu.leidenuniv.AuToBIAdapter.DataSetBuilder(params)
        except Py4JJavaError:
            raise ValueError("Cannot read arguments. Could not parse Textgrid or Wav.")

    def with_default_features(self, name: str) -> Self:
        try:
            self._object.withDefaultFeatures(name)
        except Py4JJavaError:
            raise ValueError(f"Cannot add feature {name}, not found")
        return self

    def with_feature(self, name: str) -> Self:
        try:
            self._object.withFeature(name)
        except Py4JJavaError:
            raise ValueError(f"Cannot add feature {name}, not found")
        return self

    def with_features(self, names: List[str]) -> Self:
        _name = "\t".join(names)
        try:
            self._object.withFeatures(_name)
        except Py4JJavaError:
            raise ValueError(f"Cannot add features {names}, not found")
        return self

    @takes(Path, converter=to_resolved_path_str)
    def write_csv(self, filename: str):
        self._object.writeCSV(filename)

    @takes(Path, converter=to_resolved_path_str)
    def write_arff(self, filename: str):
        self._object.writeARFF(filename)

    @takes(Path, converter=to_resolved_path_str)
    def write_liblinear(self, filename: str):
        self._object.writeLibLinear(filename)

    def build_pandas(self) -> pd.DataFrame:
        with TemporaryDirectory() as dir:
            tempfile = Path(dir).joinpath("out.csv")
            self.write_csv(tempfile)
            return pd.read_csv(tempfile, delimiter=",")

File: adapter/src/main/scala/edu/leidenuniv/AuToBIAdapter/DataSetBuilder.scala
package edu.leidenuniv.AuToBIAdapter

import edu.cuny.qc.speech.AuToBI.core.{AuToBIParameters, FeatureSet, Word}
import edu.leidenuniv.AuToBIAdapter.core.{
  fromArgsString,
  FeatureExtractors,
  FeatureSets,
  InputFiles,
}
import java.util.List as JavaList

class DataSetBuilder(val featureSet: FeatureSet):
  def this(dataPoints: JavaList[Word]) =
    this(FeatureSet())
    featureSet.setDataPoints(dataPoints)

  def this(params: AuToBIParameters) =
    this(InputFiles.getData(params).get)

  def this(params: String) =
    this((new AuToBIParameters()).fromArgsString(params))

  def withDefaultFeatures(name: String): this.type =
    FeatureSets.getFeatureSet(name) match
      case None    =>
        throw new IllegalArgumentException(s"Cannot add features $name, not found")
      case Some(x) => x.getRequiredFeatures.forEach(featureSet.insertRequiredFeature(_))
    this

  def withFeature(name: String): this.type =
    if FeatureExtractors.checkFeaturePossible(name) then
      featureSet.insertRequiredFeature(name)
    else
      throw new IllegalArgumentException(
        s"Cannot add feature $name, no extractor available",
      )
    this

  def withFeatures(names: List[String]): this.type =
    names.foreach(withFeature)
    this

  def withFeatures(names: String): this.type =
    withFeatures(names.split('\t').toList)

  private def build(): Unit =
    FeatureExtractors(featureSet).extractFeatures()
    featureSet.constructFeatures()

  def writeCSV(filename: String): Unit =
    build()
    featureSet.writeCSVFile(filename)

  def writeARFF(filename: String, name: String): Unit =
    build()
    featureSet.writeArff(filename, name)

  def writeLibLinear(filename: String): Unit =
    build()
    featureSet.writeLibLinear(filename)
